### PR TITLE
Grab design levels

### LINKED ISSUE #41: model matrix and grab_psiFun can drop columns when subsetting data by rows
This is a particularly important issue when using `grab_psiFUN()` within an eeFun.

What (i think) happens is that when passing a subset (by row) of the data to `grab_psiFun()`, then its underlying call to `model.matrix()` can drop some levels of the factor variables if they are not present in the data's subset. This is an issue when the model parameters have fixed length (`p`): when the non-present factor levels are dropped from model.matrix such that `ncol(xmatrix)<p`, then `xmatrix %*% fixed_effects` will throw an error.

I'll have PR soon that hopefully fixes this.

### PR DESCRIPTION
**Pull request by BarkleyBG**
* bumps to v1.0.5
* Fixes #41 
* Added `grab_design_levels` to find all factor-levels in a design matrix. This output can then be passed to the `xlev` arg in `model.matrix` or more importantly through `grab_psiFun` for the models. This fixes issue 41.              
     - Notably, if this seems to work well, then perhaps `grab_design_levels` could be added to the body of `grab_psiFun` to automatically handle this issue.
* Edited `binomial_integrand` to hopefully handle singleton groups better. It will now attempt to coerce back to a one-row matrix before calling `apply`.

### CODE REVIEW

**Comment by BarkleyBG** [comment]
I think that `grab_design_levels` _could_ be  added to each of the `grab_psiFun`s. However, I did not do it here, because i didn't do any deep testing on whether it would work or not. I think this would be necessary:

- find a case where `grab_psiFun` breaks without specifying the design matrix levels
    - should be easy enough to find a model with lots of factors, and a subset of the data that doesn't have all those factors. And then run the entirety of a eeFun through `m_estimate`.
- Add the above into a test, and then see if adding `grab_design_levels` into the body of `grab_psiFun` will fix it

**Comment by bsaul** [comment]
Brian, 
Thanks for finding this issue and finding a solution.
2 things:
-can you specify the namespace of `model.frame` as in `stats::model.frame`? See error [here](https://travis-ci.org/bsaul/geex/builds/345657862?utm_source=github_status&utm_medium=notification).
- can you submit this pull request to the develop branch?

**Comment by BarkleyBG** [comment]
gah, sorry about namespace issue. On the bright side, I learned about `devtools::check_man()`.

- do you want me to change version back to 1.0.4, or use 1.0.4.9000 or something?

Suggestion: perhaps `grab_psiFun()` can take a `use_levels` argument. When `use_levels=TRUE`, then this would pass along  `xlev=grab_design_levels(model)` within `grab_psiFun()`. Perhaps default to FALSE for the time being.

**Comment by BarkleyBG** [comment]
I changed this PR to base branch develop, not master. I think GH allows the repo owner to do this now?

**Comment by bsaul** [comment]
There's a strange bug in the `\dontrun{}` within the `grab_design_levels` documentation:
image

It's not obvious to me what the problem is. Any ideas?

**Comment by BarkleyBG** [comment]
I'm 99% certain it's me screwing up `\donotrun{}` instead of `\dontrun{}`.


https://cran.r-project.org/doc/manuals/R-exts.html#Documenting-functions
>  You can use \dontrun{} for text that should only be shown

**Comment by BarkleyBG** [comment]
I had fixed the code, but had forgotten to re-roxygenise. I ran `devtools::document()` and pushed.

File: R/grab_psiFUN_funs.R [hunk 1]
```diff
@@ -208,11 +208,15 @@ objFun_glmerMod_binomial <- function(parms, response, xmatrix, linkinv)
...
-  }
+  # if(class(xmatrix) != 'matrix'){
+  #   xmatrix <- as.matrix(xmatrix)
+  #   warning(paste("xmatrix was not a matrix but now it has dims", dim(xmatrix)))
+  # }
   pr  <- linkinv( drop(outer(xmatrix %*% parms[-length(parms)], b, '+') ) )
   hh  <- stats::dbinom(response, 1, prob = pr)
+  if (!is.matrix(hh)){
+    hh <- as.matrix(hh, ncol=1, nrow=length(hh))
+  }
```
Comment (BarkleyBG) [inline R/grab_psiFUN_funs.R:219]: these changes to `binomial_integrand` should fix #41

**Pull request merged**